src/shmem_team.c

```diff
@@ -0,0 +1,460 @@
+    if (NULL == shmem_internal_psync_pool) return -1;
+
+    for (long i = 0; i < psync_len; i++) {
+        shmem_internal_psync_pool[i] = SHMEM_SYNC_VALUE;
+    }
+
+    /* Convenience pointer to the group-3 pSync array (for barriers and syncs): */
+    shmem_internal_psync_barrier_pool = &shmem_internal_psync_pool[PSYNC_CHUNK_SIZE * max_teams];
+
+    if (max_teams > N_PSYNC_BYTES * CHAR_BIT) {
+        RAISE_ERROR_MSG("Requested %ld teams, but only %d are supported\n",
+                         max_teams, N_PSYNC_BYTES * CHAR_BIT);
+    }
+
+    psync_pool_avail = shmem_internal_shmalloc(2 * N_PSYNC_BYTES);
+    psync_pool_avail_reduced = &psync_pool_avail[N_PSYNC_BYTES];
+
+    /* Initialize the psync bits to 1, making all slots available: */
+    for (size_t i = 0; i < N_PSYNC_BYTES; i++) {
+        psync_pool_avail[i] = ~0;
+    }
+
+    /* Set the bits for SHMEM_TEAM_WORLD and SHMEM_TEAM_SHARED to 0: */
+    shmem_internal_bit_clear(psync_pool_avail, N_PSYNC_BYTES, SHMEMX_TEAM_WORLD_INDEX);
+    shmem_internal_bit_clear(psync_pool_avail, N_PSYNC_BYTES, SHMEMX_TEAM_SHARED_INDEX);
+
+    for (size_t i = 0; i < max_teams; i++) {
+        shmem_internal_team_pool[i] = NULL;
+    }
+    shmem_internal_team_pool[SHMEMX_TEAM_WORLD_INDEX] = &shmem_internal_team_world;
+    shmem_internal_team_pool[SHMEMX_TEAM_SHARED_INDEX] = &shmem_internal_team_shared;
+
+    return 0;
+}
+
+void shmem_internal_team_fini(void)
+{
+    /* Destroy all undestroyed teams */
+    for (long i = 0; i < shmem_internal_params.TEAMS_MAX; i++) {
+        if (shmem_internal_team_pool[i] != NULL)
+            shmem_internal_team_destroy(shmem_internal_team_pool[i]);
+    }
+
+    free(shmem_internal_team_pool);
+    shmem_internal_free(shmem_internal_psync_pool);
+    shmem_internal_free(psync_pool_avail);
+
+    return;
+}
+
+int shmem_internal_team_my_pe(shmem_internal_team_t *team)
+{
+    if (team == SHMEMX_TEAM_INVALID)
+        return -1;
+    else
+        return ((shmem_internal_team_t *)team)->my_pe;
+}
+
+int shmem_internal_team_n_pes(shmem_internal_team_t *team)
+{
+    if (team == SHMEMX_TEAM_INVALID)
+        return -1;
+    else
+        return ((shmem_internal_team_t *)team)->size;
+}
+
+void shmem_internal_team_get_config(shmem_internal_team_t *team, shmemx_team_config_t *config)
+{
+    shmem_internal_team_t *myteam = (shmem_internal_team_t *)team;
+    *config = myteam->config;
+    return;
+}
+
+int shmem_internal_team_translate_pe(shmem_internal_team_t *src_team, int src_pe,
+                                     shmem_internal_team_t *dest_team)
+{
+    int src_pe_world, dest_pe = -1;
+
+    if (src_team == SHMEMX_TEAM_INVALID || dest_team == SHMEMX_TEAM_INVALID)
+        return -1;
+
+    if (src_pe > src_team->size)
+        return -1;
+
+    src_pe_world = src_team->start + src_pe * src_team->stride;
+
+    if (src_pe_world < src_team->start || src_pe_world >= shmem_internal_num_pes)
```

Comment:
Shouldn't it be impossible to get `src_pe_world >= shmem_internal_num_pes`?  If so, this should be an assert that leads to failure, not `return -1`.

Reply:
Yeah, I think we're in undefined behavior here.  The same is true for `src_pe_world < src_team->start` right?  Do we want this to be an `assert` or an `assertp`?